src/System.Diagnostics.PerformanceCounter/src/System/Diagnostics/PerformanceCounterLib.cs

```diff
@@ -1182,8 +1203,21 @@ internal static void RegisterCategory(string categoryName, PerformanceCounterCat
                     CreateSymbolFile(creationData);
                     RegisterFiles(IniFilePath, false);
                 }
-                CloseAllTables();
-                CloseAllLibraries();
+                /*
+                  Wait some seconds for publication, sometimes registry modification is not immediatly visibile to all.
+                  (https://msdn.microsoft.com/en-us/library/cs38wsc4%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396)
+                  Similar pattern of GetStringTable()  "int waitRetries = 14;   //((2^13)-1)*10ms == approximately 1.4mins"
+                 */
+                bool isPublished = false;
+                DateTime now = DateTime.UtcNow;
+                while ((DateTime.UtcNow.Subtract(now).TotalSeconds < 10))
```

Comment:
I feel reasonably good about adding locks here, but I don't think that we should add a 10 second retry here.  If you want to add a couple of retries, that might be OK.  I also don't think that I would throw an exception either.  The only real thing that can be done in the case of the thrown exception is to wait longer, and so you'd expect the user of performance counters to do that when attempting to get the counter - users already have to do this, right?

Reply:
I see your point - It's just not clear to me what the right number of retries or the correct amount of time is here.  That's why I would guess that the desktop implementation does not have this type of logic and instead depends upon the caller to decide what to do (not necessarily better, but more flexible).

@weshaggard, do you know if there is any precedent in managed APIs that wrap Windows OS APIs where delay is an issue (such as this registry publication delay)?  Do we generally take the approach that callers need to do retries, or do we have a pattern that we've used before to attempt to solve this problem?

Reply:
@stephentoub could answer this. I thought there are some examples (maybe in IO?) but I can't find one specifically so that may not be the case.

Reply:
Thanks @danmosemsft.  @stephentoub do you have any thoughts on this?

Reply:
I don't think we should add retries here.  There's no way we can come up with a reasonable policy, if the underlying APIs we use to publish don't guarantee immediacy.

Reply:
Looked a bit more, as @MarcoRossignoli points out in his comment, there are already retries in this library. (I searched for 'retries'). If we accept that the native API has some intrinsic delays and offers no wait to wait efficiently then I guess it is appropriate for this library to do some retries. What we should not do is retries to work around defects in our own implementation, of course.

The existing retries on PerformanceCounterLib.GetData say that they will wait up to 10 minutes! That seems very high.

Reply:
As i heve written above IMHO with this underlying APIs we can only "mitigate" this issue, parallel unit testing is borderline, i got the idea from other place on this namespace as @danmosemsft said. I agree with you @stephentoub, there isn't a "definitive policy" to solve this issue, we can try with "timeout retries" style or we need to leave [assembly:CollectionBehavior(DisableTestParallelization = true)] and fix only races. By the way we can get "not found" also with this workaround if for some reason the  registry delay is high.

Reply:
ping

Reply:
@danmosemsft, @stephentoub I'm hoping that you guys can help me understand the precedent that we have in CoreFX to determine if the retries are OK here or not.

Reply:
Just to be clear, this is not retries, this is 'wait for publication'.     Previously the API was asynchronous (you set things in motion, but state update may not have happened before return), and you have made the synchronous (which gets you into the business of deciding how long to wait, and what to do when that fails.  

We ARE changing the contract (it can throw an exception now where it definitely did NOT before), so this is a compatibility break (how much do we care?)   

This is also a 'compatibility' library (it is not part of the cross platform core), and my expectation is that we are not trying to innovate here (just make existing things work).   

This suggests that we should not change things here.    What are the disadvantages of simply leaving the old behavior?

Reply:
@vancem i agree with you this is 'wait for publication' and the idea is to mitigate the issue. The scope of this PR is for first resolve some race issues and now seems ok. The second is try to remove [assembly:CollectionBehavior(DisableTestParallelization = true)] to run tests concurrently, but maybe this is not fundamental. @danmosemsft added this attribute after some tests fail due to "wait for publication" side effect and internal cleanup/load cached tables. If we don't want to change too much on this namespace we can keep the attribute.

Reply:
My recommendation is to pull out the wait loop  and do that separately.   If the change was only motivated by tests, it is easy enough to simply do that waiting in the test rather than here.    Which we choose really depends on whether our customers would be happier with or without it (but we are biased to doing nothing if we don't know).   I am suspicious that we don't know, so I would keep the semantics the same (less delta between us an .NET Desktop).   

So my recommendation is to pull it (we can put it back if we care in the future...

Reply:
ok @vancem i'll remove 'wait for publication strategy' and fix only race, thank's! @danmosemsft i think we need to keep attribute(or expose `IsPublished()` to category and update all tests if it's so important).

Reply:
Thanks @MarcoRossignoli!